src/heuristic/HeuristicMapper.cpp

```diff
@@ -597,3 +615,162 @@ void HeuristicMapper::lookahead(const std::size_t      layer,
+    target = swap.first;
+  }
+
+  if (source == middleAnc || target == middleAnc) {
+    std::clog << "FAIL: TELE " << source << " -(" << middleAnc << ")-> "
+              << target << "\n";
+    throw QMAPException("Overlap between source/target and middle "
+                        "ancillary in teleportation.");
+  }
+
+  swaps.back().emplace_back(source, target, middleAnc, qc::Teleportation);
+
+  costFixed += COST_TELEPORTATION;
+}
+
+void HeuristicMapper::Node::recalculateFixedCost(const Architecture& arch) {
+  costFixed = 0;
+  for (auto& swapNode : swaps) {
+    for (auto& swap : swapNode) {
+      if (swap.op == qc::SWAP) {
+        if (arch.bidirectional()) {
+          costFixed += COST_BIDIRECTIONAL_SWAP;
+        } else {
+          costFixed += COST_UNIDIRECTIONAL_SWAP;
+        }
+      } else if (swap.op == qc::Teleportation) {
+        costFixed += COST_TELEPORTATION;
+      }
+    }
+  }
+}
+
+void HeuristicMapper::Node::updateHeuristicCost(
+    const Architecture&         arch,
+    const TwoQubitMultiplicity& twoQubitGateMultiplicity,
+    bool                        admissibleHeuristic) {
+  costHeur = 0.;
+  done     = true;
+
+  // iterating over all virtual qubit pairs, that share a gate on the
+  // current layer
+  for (const auto& edgeMultiplicity : twoQubitGateMultiplicity) {
+    const auto& q1                   = edgeMultiplicity.first.first;
+    const auto& q2                   = edgeMultiplicity.first.second;
+    const auto& straightMultiplicity = edgeMultiplicity.second.first;
+    const auto& reverseMultiplicity  = edgeMultiplicity.second.second;
+
+    bool edgeDone = (arch.getCouplingMap().find(
+                         {static_cast<std::uint16_t>(locations.at(q1)),
+                          static_cast<std::uint16_t>(locations.at(q2))}) !=
+                         arch.getCouplingMap().end() ||
+                     arch.getCouplingMap().find(
+                         {static_cast<std::uint16_t>(locations.at(q2)),
+                          static_cast<std::uint16_t>(locations.at(q1))}) !=
+                         arch.getCouplingMap().end());
+    // only if all qubit pairs are mapped next to each other the mapping
+    // is complete
+    if (!edgeDone) {
+      done = false;
+    }
+
+    double swapCostStraight =
+        arch.distance(static_cast<std::uint16_t>(locations.at(q1)),
+                      static_cast<std::uint16_t>(locations.at(q2)));
+    double swapCostReverse =
+        arch.distance(static_cast<std::uint16_t>(locations.at(q2)),
+                      static_cast<std::uint16_t>(locations.at(q1)));
+
+    if (admissibleHeuristic) {
+      if (straightMultiplicity > 0) {
+        costHeur = std::max(costHeur, swapCostStraight);
+      }
+      if (reverseMultiplicity > 0) {
+        costHeur = std::max(costHeur, swapCostReverse);
+      }
+    } else {
+      costHeur += swapCostStraight * straightMultiplicity +
+                  swapCostReverse * reverseMultiplicity;
```

Comment:
In most cases, this will be an overestimate, right? For example, a multiplicity `{2, 0}` will only require a single `swapStraight`, right?
At the moment, this resembles the state of the `main` branch. Could this be improved somehow?

Reply:
That is true. I copied the behavior of the `main` branch exactly. I thought, that a pure code refactor PR probably should not change the fundamental functionality. In its current form the code in this PR yields exactly the same A* nodes (with exactly the same costs) just with some little performance boost.

But if you find it appropriate, we could also optimize the above already in this PR. I think the best solution would be `costHeur += std::max(swapCostStraight, swapCostReverse)`.

However, I also thought a little bit about the idea of improving the admissible heuristic using the fact that any moving qubit can save at most a number of swaps equal to the number of other (potentially) moving qubits, and am planning on making another PR on that once we discussed it. So we could also wait with the above change for that PR.

Reply:
I agree that it's probably best to keep this as a pure refactoring PR.
The follow-up PR can then improve both heuristics.

I like you suggestion for the non-admissible part. Should be as straight forward as that.

As for the admissible part, the idea makes sense, but I guess it's easier to reason over some code in that regard.